performance_manager/lib/l1_full_trip_events.py

```diff
@@ -0,0 +1,389 @@
+                event_type.merge_dataframe[column] = pandas.to_numeric(
+                    event_type.merge_dataframe[column]
+                ).astype("Int64")
+            # create hash column to be used as FullTripEvent primary key
+            # drop category columns after hash has been created
+            event_type.merge_dataframe = add_event_hash_column(
+                event_type.merge_dataframe,
+                expected_hash_columns=expected_hash_columns,
+            )[["pk_id", "hash"]]
+
+        # rename pk_id field from L0 table to L1 foreign key field name
+        event_type.merge_dataframe.rename(
+            columns={"pk_id": event_type.fk_rename_field}, inplace=True
+        )
+
+
+def merge_events(events_list: Dict[str, EventsToMerge]) -> pandas.DataFrame:
+    """
+    merge all unprocessed event types into single dataframe
+    """
+    # left merge TRIP_UPDATE stop events with moving VEHICLE_POSITION events
+    # TRIP_UPDATE predicted stop times are only helpful as supplement to
+    # VEHICLE_POSITION moving event
+    return_dataframe = events_list["moving_vp"].merge_dataframe.merge(
+        events_list["stopped_tu"].merge_dataframe, how="left", on="hash"
+    )
+    # outer join VEHICLE_POSITION stop events to capture all recorded stop
+    # events, even if they do not have associated moving event
+    return_dataframe = return_dataframe.merge(
+        events_list["stopped_vp"].merge_dataframe, how="outer", on="hash"
+    )
+
+    # merge action turns foreign key field into "float" type, this will
+    # convert foreign keys back to integer for update / insert action
+    for column in return_dataframe.columns:
+        if column == "hash":
+            continue
+        return_dataframe[column] = return_dataframe[column].astype("Int64")
+
+    # make sure all "na" types are replace as Python None for RDS
+    # insert / upate
+    return_dataframe = return_dataframe.fillna(numpy.nan).replace(
+        [numpy.nan], [None]
+    )
+
+    return return_dataframe
+
+
+def load_temp_table(
+    insert_dataframe: pandas.DataFrame, db_manager: DatabaseManager
+) -> None:
+    """
+    truncate temp loading table TempFullTripEvents and insert new
+    trip event records for update / insert with FullTripEvents table
+    """
+    db_manager.truncate_table(TempFullTripEvents)
+
+    db_manager.insert_dataframe(insert_dataframe, TempFullTripEvents.__table__)
+
+
+def update_with_new_events(
+    insert_dataframe: pandas.DataFrame, db_manager: DatabaseManager
+) -> None:
+    """
+    update FullTripEvents table if insert_dataframe has overlapping hash values
+
+    this can normally be done with and sql UPDATE-FROM query, but sqlalchemy
+    does not support that type of query with sqlite dev database
+
+    alternate approach is used here where just hashes are selected that represent
+    update records and each column is updated seperatly from insert_dataframe
+    """
```

Comment:
updated this to be a true update operation, updating each column individually so it plays nice with sqlite, postgres could probably do this as one UPDATE from SELECT statement